lib/rucio/core/replica.py

```diff
@@ -1544,21 +1590,47 @@ def delete_replicas(rse_id, files, ignore_availability=True, session=None):
-                                                  ~exists(select([1]).prefix_with("/*+ INDEX(CONTENTS CONTENTS_PK) */", dialect='oracle')).where(and_(models.DataIdentifierAssociation.scope == parent_scope,
-                                                                                                                                                      models.DataIdentifierAssociation.name == parent_name))))
-
-                tmp_parent_condition.append(and_(models.DataIdentifierAssociation.child_scope == parent_scope, models.DataIdentifierAssociation.child_name == parent_name,
-                                                 ~exists(select([1]).prefix_with("/*+ INDEX(CONTENTS CONTENTS_PK) */", dialect='oracle')).where(and_(models.DataIdentifierAssociation.scope == parent_scope,
-                                                                                                                                                     models.DataIdentifierAssociation.name == parent_name))))
-
-                did_condition.append(and_(models.DataIdentifier.scope == parent_scope, models.DataIdentifier.name == parent_name, models.DataIdentifier.is_open == False,
-                                          ~exists([1]).where(and_(models.DataIdentifierAssociation.child_scope == parent_scope, models.DataIdentifierAssociation.child_name == parent_name)),  # NOQA
-                                          ~exists([1]).where(and_(models.DataIdentifierAssociation.scope == parent_scope, models.DataIdentifierAssociation.name == parent_name))))  # NOQA
+                # Schedule removal of child file/dataset/container from the parent dataset/container
+                child_did_condition.append(
+                    and_(models.DataIdentifierAssociation.scope == parent_scope,
+                         models.DataIdentifierAssociation.name == parent_name,
+                         models.DataIdentifierAssociation.child_scope == child_scope,
+                         models.DataIdentifierAssociation.child_name == child_name))
+
+                # If the parent dataset/container becomes empty as a result of the child removal
+                # (it was the last children), metadata cleanup has to be done:
+                #
+                # 1) Schedule to remove the replicas of this empty collection
+                clt_replica_condition.append(
+                    and_(models.CollectionReplica.scope == parent_scope,
+                         models.CollectionReplica.name == parent_name,
+                         exists(select([1]).prefix_with("/*+ INDEX(DIDS DIDS_PK) */", dialect='oracle')).where(
+                             and_(models.DataIdentifier.scope == parent_scope,
+                                  models.DataIdentifier.name == parent_name,
+                                  models.DataIdentifier.is_open == False)),  # NOQA
```

Comment:
Does anyone remember why we have this ? I suspect this is the cause of this bug: https://github.com/rucio/rucio/issues/3311

Reply:
I can't remember, but thinking about this, it is probably wrong and should not filter on `is_open == False`.

Reply:
Let's continue discussing this in #3311
I think @rcarpa can have a look at this one too during his reaper2 works.